### DISCUSSION TITLE
Support multiple callbacks for `router.beforePopState()`

### DISCUSSION THREAD

**Discussion by ericbiewener** [vercel/next.js#34835] [Ideas]
### Describe the feature you'd like to request

Currently, `router.beforePopState()` only supports [a single callback function](https://github.com/vercel/next.js/blob/v12.1.1-canary.3/packages/next/shared/lib/router/router.ts#L1622). Subsequent calls to that function will wipe out any previously supplied callback functions. Perhaps this is intentional, but I'd prefer it to support a variable number of callbacks. I think this would be both functionally and intuitively superior.

My use case is simply that multiple components currently active in the hierarchy may wish to supply their own callback.

### Describe the solution you'd like

1. `beforePopState()` should simply push the supplied callback on to an array
2. It would return a cleanup function so that callees can remove the callback when unmounting (e.g. as a return value from a `useEffect` hook
3. If _any_ of the supplied callbacks return `false`, then the state change would be prevented, thus continuing to support the current ability to prevent a state change.

### Describe alternatives you've considered

I've currently resorted to providing a proxied version of the router via context in my app so that I can augment the functionality in this way.

**Comment by ericbiewener** [comment]
Happy to open a PR if this is desired.

**Comment by Sh031224** [comment]
I think there has to be a way.
